sos/collector/transports/saltstack.py

```diff
@@ -0,0 +1,173 @@
+from sos.utilities import is_executable
+
+TIMEOUT_DEFAULT = 300
+
+
+class SaltStackMasterTransport(RemoteTransport):
+    """
+    A transport for collect that leverages SaltStack's Master Pub/Sub
+    functionality to send commands to minions.
+
+    This transport will by default assume the use cmd.shell module to
+    execute commands on the minions.
+    """
+
+    name = 'saltstack'
+
+    def _convert_output_json(self, json_output):
+        return list(json.loads(json_output).values())[0]
+
+    def run_command(
+            self, cmd, timeout=180, need_root=False, env=None, get_pty=False):
+        """
+        Run a command on the remote host using SaltStack Master.
+        If the output is json, convert it to a string.
+        """
+        ret = super(SaltStackMasterTransport, self).run_command(
+            cmd, timeout, need_root, env, get_pty)
+        with contextlib.suppress(Exception):
+            ret['output'] = self._convert_output_json(ret['output'])
+        return ret
+
+    def _salt_retrieve_file(
+            self, node, file, failure_message,
+            timeout=180, need_root=False, env=None, get_pty=False):
+        """
+        Execute cp.push on the remote host using SaltStack Master
+        """
+        command = "salt %s cp.push %s" % (node, file)
+        command = command.split()
+        cmd = Popen(
+            command, stdout=PIPE, stderr=PIPE)
+        out, err = cmd.communicate()
+        err = err.decode('utf-8')
+        return {'status': 0} if out and not err else {'status': 1}
+
+    @property
+    def connected(self):
+        """Check if the remote host is responding using SaltStack Master."""
+        up = self.run_command("echo Connected", timeout=10)
+        return up['status'] == 0
+
+    def _check_for_saltstack(self, password=None):
+        """Checks to see if the local system supported SaltStack Master.
+
+        This check relies on feedback from the salt binary. The command being
+        run should always generate stderr output, but depending on what that
+        output reads we can determine if SaltStack Master is supported or not.
+
+        For our purposes, a host that does not support SaltStack Master is not
+        able to run sos-collector.
+
+        Returns
+            True if SaltStack Master is supported, else raise Exception
+        """
+
+        def _check_for_salt_master(command, failure_message):
+            """Check for various salt master binaries on the local system."""
+            cmd = Popen(
+                command, stdout=PIPE, stderr=PIPE)
+            out, err = cmd.communicate()
+            err = err.decode('utf-8')
+            if 'command not found' in err or 'Usage:' in err:
+                raise ConnectionException(failure_message)
+
+        salt_cmd = ['salt', '--version']
+        salt_run_cmd = ['salt-run', 'manage.status']
+        _check_for_salt_master(
+            salt_cmd, "The salt binary is not available on this system")
```

Comment:
Mostly redundant, at least for `salt`. I'll remove this and rely on **_connect()**'s **is_executable('salt')**

The additional check for `salt-run` confirms the **salt-master** service is responding local to the sos collector. Otherwise the master may not be in a functional state to send commands to minions. Unless you have a preferred placement for this, I'll leave the **_check_for_salt_master()** verifying the service responds.

Reply:
Going to leave this checking `salt-run manage.status` for the time being as its possible the service active in terms of systemd, but potentially not functional in which case I think the current check results in a more thorough test.